### PR TITLE
Fix incorrect return of null values in L2C

### PR DESCRIPTION
**Pull request by AlexSmerw**
Bug fix for issue [7735](https://github.com/doctrine/orm/issues/7735) for doctrine 2.6

### CODE REVIEW

**Review by lcobucci — COMMENTED** [review]
@AlexSmerw thanks for your contribution! 

This does look sane from a very quick look but could you please add a functional that reproduces the issue and proves that the fix solves it?

**Comment by AlexSmerw** [comment]
> @AlexSmerw thanks for your contribution!
> 
> This does look sane from a very quick look but could you please add a functional that reproduces the issue and proves that the fix solves it?

This is not enough -  [description with link on code](https://github.com/doctrine/orm/issues/7735#issuecomment-503555655) ?

**Comment by lcobucci** [comment]
Not really, we need a phpunit test that gets integrated to the codebase, which also ensures that this problem will never happen again. 

You can find examples on https://github.com/doctrine/doctrine2/tree/388afb46d0cb3ed0c51332e8df0de9e942c2690b/tests/Doctrine/Tests/ORM/Functional/Ticket

**Comment by AlexSmerw** [comment]
> Not really, we need a phpunit test that gets integrated to the codebase, which also ensures that this problem will never happen again.
> 
> You can find examples on https://github.com/doctrine/doctrine2/tree/388afb46d0cb3ed0c51332e8df0de9e942c2690b/tests/Doctrine/Tests/ORM/Functional/Ticket

I have added unitTest in my branch. If you use this test **Issue7735Test** without my edits, you will get Exception `TypeError: Return value of Doctrine\Tests\Models\Issue7735\Engine::getModel() must be of the type string, null returned`. Look, I set 1000000 iterations for getting this bug. I think this count depends on you PC

File: tests/Doctrine/Tests/OrmFunctionalTestCase.php [hunk 1]
```diff
@@ -724,7 +732,10 @@ protected function _getEntityManager(
 
             $cacheConfig    = new CacheConfiguration();
             $cache          = $this->getSharedSecondLevelCacheDriverImpl();
-            $factory        = new DefaultCacheFactory($cacheConfig->getRegionsConfiguration(), $cache);
+
+            $regionConfig = $cacheConfig->getRegionsConfiguration();
+            $regionConfig->setDefaultLifetime(static::DEFAULT_CACHE_REGION_LIFE_TIME);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/OrmFunctionalTestCase.php]: Is this change really required for the test?

**Review by lcobucci — CHANGES_REQUESTED** [review]
Thank you very much for adding the test, I've got a tiny adjustment request: can you please put the models in the same file as the test and remove all attributes which might be irrelevant to reproducible the issue?

File: tests/Doctrine/Tests/OrmFunctionalTestCase.php (hunk 1 above)
Comment (AlexSmerw) [reply tests/Doctrine/Tests/OrmFunctionalTestCase.php]: For reproducing this problem,  I set second level cache time limit  = 1 second. This code need for this.

Comment (lcobucci) [reply tests/Doctrine/Tests/OrmFunctionalTestCase.php]: Sorry, what I meant to say was: can't we keep this contained to the functional test only? This is now changing the default for every test in the suite, which might create undesired side-effects 😁

**Comment by AlexSmerw** [comment]
> Thank you very much for adding the test, I've got a tiny adjustment request: can you please put the models in the same file as the test and remove all attributes which might be irrelevant to reproducible the issue? 
Hi! I have done this.

**Comment by lcobucci** [comment]
@AlexSmerw it seems like we still have separate files for each class... Check this example: https://github.com/doctrine/orm/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/tests/Doctrine/Tests/ORM/Functional/Ticket/GH6362Test.php

File: tests/Doctrine/Tests/OrmFunctionalTestCase.php (hunk 1 above)
Comment (AlexSmerw) [reply tests/Doctrine/Tests/OrmFunctionalTestCase.php]: > Sorry, what I meant to say was: can't we keep this contained to the functional test only? This is now changing the default for every test in the suite, which might create undesired side-effects grin

Done

**Comment by AlexSmerw** [comment]
> @AlexSmerw it seems like we still have separate files for each class... Check this example: https://github.com/doctrine/orm/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/tests/Doctrine/Tests/ORM/Functional/Ticket/GH6362Test.php

Oh! It's Evil!

**Comment by AlexSmerw** [comment]
> @AlexSmerw it seems like we still have separate files for each class... Check this example: https://github.com/doctrine/orm/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/tests/Doctrine/Tests/ORM/Functional/Ticket/GH6362Test.php

Done. Check it, please.

**Comment by lcobucci** [comment]
@AlexSmerw I'll take a look at it ASAP, thanks for applying the changes!

**Comment by AlexSmerw** [comment]
> @AlexSmerw I'll take a look at it ASAP, thanks for applying the changes!
Hi!
My request still with label Missing Tests. Is it ok?
I worry about its fate :)

**Comment by greg0ire** [comment]
Please associate your commit email address with your Github account, or change the email in your commits to an address already associated with it.

File: lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php [hunk 2]
```diff
@@ -452,12 +452,12 @@ public function loadById(array $identifier, $entity = null)
...
             }
 
-            if (($entity = $this->hydrator->loadCacheEntry($class, $cacheKey, $cacheEntry, $entity)) !== null) {
+            if (($cachedEntity = $this->hydrator->loadCacheEntry($class, $cacheKey, $cacheEntry, $entity)) !== null) {
                 if ($this->cacheLogger) {
                     $this->cacheLogger->entityCacheHit($this->regionName, $cacheKey);
                 }
 
-                return $entity;
+                return $cachedEntity;
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php]: A good commit message might be:

```
Avoid reusing variable name

The same variable name is used below, and that causes a bug etc.
Fixes #issueNumber
```

Comment (lcobucci) [reply lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php]: Commit message updated :+1:

**Review by lcobucci — DISMISSED** [review]
@AlexSmerw I've managed to clean up your test and remove the need for the cache TTL configuration and the `for` loop.

Thank you very much for fixing this bug!

**Review by lcobucci — APPROVED** [review]
